File: client/packages/socket_rtsp/lib/rtp_packet.dart
import 'dart:typed_data';

const int HEADER_SIZE = 12;

class RtpPacket {
  final Uint8List header = Uint8List(HEADER_SIZE);
  Uint8List? payload;

  RtpPacket();

  void encode(int version, bool padding, bool extension, int cc, int seqnum,
      bool marker, int pt, int ssrc, Uint8List payload) {
    final timestamp = DateTime.now().millisecondsSinceEpoch;

    header[0] = (3 << 6) & 0xFF; // version bits
    header[0] |= (padding ? 1 << 5 : 0);
    header[0] |= (extension ? 1 << 4 : 0);
    header[0] |= cc & 0xFF;

    header[1] = (marker ? 1 << 7 : 0) & 0xFF;
    header[1] |= pt & 0xFF;

    header[2] = (seqnum >> 8) & 0xFF;
    header[3] = seqnum & 0xFF;

    header[4] = (timestamp >> 24) & 0xFF;
    header[5] = (timestamp >> 16) & 0xFF;
    header[6] = (timestamp >> 8) & 0xFF;
    header[7] = timestamp & 0xFF;

    this.payload = payload;
  }

  void decode(Uint8List byteStream) {
    header.setAll(0, byteStream.sublist(0, HEADER_SIZE));
    payload = byteStream.sublist(HEADER_SIZE);
  }

  int version() {
    return (header[0] >> 6) & 3;
  }

  int seqNum() {
    return (header[2] << 8) | header[3];
  }

  int timestamp() {
    return (header[4] << 24) | (header[5] << 16) | (header[6] << 8) | header[7];
  }

  int payloadType() {
    return header[1] & 127;
  }

  Uint8List getPayload() {
    return payload ?? Uint8List(0);
  }

  Uint8List getPacket() {
    return Uint8List.fromList([...header, ...getPayload()]);
  }
}

File: client/hls_test/rtp_packet.py
import sys
from time import time
HEADER_SIZE = 12

class RtpPacket:
	header = bytearray(HEADER_SIZE)

	def __init__(self):
		pass

	def encode(self, version, padding, extension, cc, seqnum, marker, pt, ssrc, payload):
		"""Encode the RTP packet with header fields and payload."""
		timestamp = int(time())
		header = bytearray(HEADER_SIZE)
		#--------------
		# TO COMPLETE
		#--------------
		# Fill the header bytearray with RTP header fields

		header[0] = 0
		header[0] = header[0] | (0b11 << 6)  & 0xFF # version bits
		header[0] = header[0] | (padding << 5)  & 0xFF
		header[0] = header[0] | (extension << 4)  & 0xFF
		header[0] = header[0] | cc & 0xFF

		header[1] = 0
		header[1] = header[1] | (marker << 7) & 0xFF
		header[1] = header[1] | 26 # payload type field (PT): MJPEG

		header[2] = (seqnum >> 8) & 0xFF
		header[3] = seqnum & 0xFF

		t = int(time())

		header[4] = (t >> 24) & 0xFF
		header[5] = (t >> 16 ) & 0xFF
		header[6] = (t >> 8) & 0xFF
		header[7] = t & 0xFF

		# Get the payload from the argument
		self.payload = payload
		self.header = header

	def decode(self, byteStream):
		"""Decode the RTP packet."""
		self.header = bytearray(byteStream[:HEADER_SIZE])
		self.payload = byteStream[HEADER_SIZE:]

	def version(self):
		"""Return RTP version."""
		return int(self.header[0] >> 6)

	def seqNum(self):
		"""Return sequence (frame) number."""
		seqNum = self.header[2] << 8 | self.header[3]
		return int(seqNum)

	def timestamp(self):
		"""Return timestamp."""
		timestamp = self.header[4] << 24 | self.header[5] << 16 | self.header[6] << 8 | self.header[7]
		return int(timestamp)

	def payloadType(self):
		"""Return payload type."""
		pt = self.header[1] & 127
		return int(pt)

	def getPayload(self):
		"""Return payload."""
		return self.payload

	def getPacket(self):
		"""Return RTP packet."""
		print(len(self.payload))
		return self.header + self.payload
